plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/ServerPortExtensionTest.java

```diff
@@ -0,0 +1,202 @@
+
+  @Before
+  public void setUp() {
+    // TODO(chanseok): use ShellTestResource (see AccountPanelTest) after fixing
+    // https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/771.
+    // (Remove shell.dispose() in tearDown() too.)
+    shell = new Shell(Display.getDefault());
+    shell.setVisible(true);
+    portExtension = new ServerPortExtension();
+    portExtension.createControl(UI_POSITION.BOTTOM, shell);
+    portExtension.setServerWorkingCopy(server);
+  }
+
+  @After
+  public void tearDown() {
+    shell.dispose();
+  }
+
+  @Test
+  public void testCreateControl() {
+    assertNotNull(portExtension.portLabel);
+    assertNotNull(portExtension.portText);
+    assertFalse(portExtension.portLabel.isVisible());
+    assertFalse(portExtension.portText.isVisible());
+  }
+
+  @Test
+  public void testDefaultPort() {
+    assertEquals("8080", portExtension.portText.getText());
+  }
+
+  @Test
+  public void testPortText_allowEmpty() {
+    portExtension.portText.setText("");
+    assertTrue(portExtension.portText.getText().isEmpty());
+  }
+
+  @Test
+  public void testPortText_preventNonNumericCharacters() {
+    portExtension.portText.setText("80A80");
+    assertEquals("8080", portExtension.portText.getText());
+
+    portExtension.portText.setText("80 clipboard pasted 80");
+    assertEquals("8080", portExtension.portText.getText());
+  }
+
+  @Test
+  public void testPortText_numericCharacters() {
+    portExtension.portText.setText("808");
+    assertEquals("808", portExtension.portText.getText());
+
+    portExtension.portText.setText("8108");
+    assertEquals("8108", portExtension.portText.getText());
+  }
+
+  @Test
+  public void testPortText_allowArbitrarilyLargeNumbers() {
+    portExtension.portText.setText("80808");
+    assertEquals("80808", portExtension.portText.getText());
+
+    portExtension.portText.setText("480808");
+    assertEquals("480808", portExtension.portText.getText());
+
+    portExtension.portText.setText("9480808");
+    assertEquals("9480808", portExtension.portText.getText());
+  }
+
+  @Test
+  public void testAttributeSetOnFocusLost() {
+    portExtension.portText.setText("12345");
+    assertEquals("12345", portExtension.portText.getText());
+
+    portExtension.portText.notifyListeners(SWT.FocusOut, null /* event */);
+    verify(server).setAttribute("appEngineDevServerPort", "12345");
+
+    portExtension.portText.setText("54321");
+    assertEquals("54321", portExtension.portText.getText());
+
+    portExtension.portText.notifyListeners(SWT.FocusOut, null /* event */);
+    verify(server).setAttribute("appEngineDevServerPort", "54321");
+  }
+
+  @Test
+  public void testPortText_limitMinimumPortOnFocusLost() {
+    portExtension.portText.setText("0");
+    assertEquals("0", portExtension.portText.getText());
+
+    portExtension.portText.notifyListeners(SWT.FocusOut, null /* event */);
+    assertEquals("1", portExtension.portText.getText());
```

Comment:
This one definitely needs to change because in TCP port 0 means "pick an available port", not use port 1 (which isn't necessarily available as non-root on Unix). This case we should just treat it the same as an empty port.